gr-soapy/lib/block_impl.cc

```diff
@@ -0,0 +1,671 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2021 Jeff Long
+ * Copyright 2018-2021 Libre Space Foundation <http://libre.space>
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "block_impl.h"
+#include <gnuradio/io_signature.h>
+#include <SoapySDR/Formats.h>
+#include <cmath>
+#include <numeric>
+
+namespace gr {
+namespace soapy {
+
+const pmt::pmt_t CMD_CHAN_KEY = pmt::mp("chan");
+const pmt::pmt_t CMD_FREQ_KEY = pmt::mp("freq");
+const pmt::pmt_t CMD_GAIN_KEY = pmt::mp("gain");
+const pmt::pmt_t CMD_ANTENNA_KEY = pmt::mp("ant");
+const pmt::pmt_t CMD_RATE_KEY = pmt::mp("rate");
+const pmt::pmt_t CMD_BW_KEY = pmt::mp("bw");
+
+/*
+ * The private constructor
+ */
+block_impl::block_impl(int direction,
+                       const std::string& device,
+                       const std::string& type,
+                       size_t nchan,
+                       const std::string& dev_args,
+                       const std::string& stream_args,
+                       const std::vector<std::string>& tune_args,
+                       const std::vector<std::string>& other_settings)
+    : d_direction(direction),
+      d_nchan(nchan),
+      d_stream_args(stream_args),
+      d_channels(nchan)
+{
+    // Channel list
+    std::iota(std::begin(d_channels), std::end(d_channels), 0);
+
+    // Must be 1 or nchan.
+    if (nchan != tune_args.size() && tune_args.size() != 1) {
+        throw std::invalid_argument(name() +
+                                    ": Wrong number of channels and tune arguments");
+    }
+
+    // Must be 1 or nchan.
+    if (nchan != other_settings.size() && other_settings.size() != 1) {
+        throw std::invalid_argument(name() + ": Wrong number of channels and settings");
+    }
+
+    // Convert type to Soapy type.
+    if (type == "fc32") {
+        d_soapy_type = SOAPY_SDR_CF32;
+    } else if (type == "sc16") {
+        d_soapy_type = SOAPY_SDR_CS16;
+    } else if (type == "sc8") {
+        d_soapy_type = SOAPY_SDR_CS8;
+    } else {
+        throw std::invalid_argument(name() + ": Invalid IO type");
+    }
+
+    SoapySDR::Kwargs dev_kwargs = SoapySDR::KwargsFromString(device + ", " + dev_args);
+    d_device = SoapySDR::Device::make(dev_kwargs);
+
+    /* Some of the arguments maybe device settings, so we need to re-apply them */
+    SoapySDR::ArgInfoList supported_settings = d_device->getSettingInfo();
+    for (const auto& arg : dev_kwargs) {
+        for (const auto& supported : supported_settings) {
+            if (supported.key == arg.first) {
+                d_device->writeSetting(arg.first, arg.second);
+            }
+        }
+    }
+
+    if (d_nchan > d_device->getNumChannels(d_direction)) {
+        throw std::invalid_argument(
```

Comment:
This will leak the `d_device`. Same with the other exceptions in this constructor.

I've not looked at the Soapy API, but if it doesn't just work with standard destructors, then maybe worth wrapping `d_device` in essentially a smart pointer that does `unmake()`?

That way RAII will work as intended.

Reply:
Using manual cleanup for now. I like the idea of a smart pointer with a custom deleter, but don't want to propagate that through the code at the moment.

Reply:
Of course, a bunch of soapysdr calls in the constructor could also throw, so I'll have to clean this up sometime. We're usually saved by the fact that the program ends and device resources are usually deallocated at that point. But some long debugging sessions could be avoided if "usually" != "always".

Reply:
Could leave a TODO, but seems to me like it should be worth it "while we're in here".

Reply:
device is a shared_ptr with a custom deleter. stream is an opaque type, so I left it as-is.

Reply:
Changed to unique_ptr.